File: Code_Help_By_Babbar/12_ Binary_Search_Explained/05_pivot.cpp
#include <iostream>
using namespace std;

int pivot(int arr[], int size)
{
    int startIdx = 0;
    int endIdx = size;
    int midIdx = startIdx + (endIdx - startIdx) / 2;

    while (startIdx < endIdx)
    {
        if (arr[midIdx] >= arr[0])
        {
            startIdx = midIdx + 1;
        }
        else
        {
            endIdx = midIdx;
        }
        midIdx = startIdx + (endIdx - startIdx) / 2;
    }
    return arr[startIdx];
}

int main()
{
    int arr[] = {6, 7, 8, 9, 1, 2, 3, 4, 5};
    int size = sizeof(arr) / sizeof(arr[0]);

    cout << "Pivot value : " << pivot(arr, size) << endl;
}

File: Topics_to_Focus_On_JS/[ 10 ] Searching/02_Binary_search/05_pivot.js
function pivot(arr) {
  let startIdx = 0;
  let endIdx = arr.length;
  let midIdx = Math.floor((startIdx + endIdx) / 2);

  while (startIdx < endIdx) {
    if (arr[midIdx] >= arr[0]) {
      startIdx = midIdx + 1;
    } else {
      endIdx = midIdx;
    }
    midIdx = Math.floor((startIdx + endIdx) / 2);
  }

  return arr[startIdx];
}

const arr = [6, 7, 8, 9, 1, 2, 3, 4, 5];
const res = pivot(arr);
console.log(res);
